src/transformers/cache_utils.py

```diff
@@ -359,11 +359,15 @@ class DynamicCache(Cache):
         ```
     """
 
-    def __init__(self, _distributed_cache_data: Optional[Iterable] = None) -> None:
+    def __init__(self, _distributed_cache_data: Optional[Iterable] = None, num_layers: Optional[int] = None) -> None:
         super().__init__()
         self._seen_tokens = 0  # Used in `generate` to keep tally of how many tokens the cache has seen
-        self.key_cache: List[torch.Tensor] = []
-        self.value_cache: List[torch.Tensor] = []
+        if num_layers is None:
+            self.key_cache: List[torch.Tensor] = []
+            self.value_cache: List[torch.Tensor] = []
+        else:
+            self.key_cache = [torch.tensor([]) for _ in range(num_layers)]
+            self.value_cache = [torch.tensor([]) for _ in range(num_layers)]
```

Comment:
why don't we always init like this?

Reply:
We need to know how many layers we want to do this for.

Reply:
`DynamicCache` has lazy tensor init, and export needs eager tensor init :D 

It's similar to the issue we have with TP (should be lazy) vs torch.compile (should be eager)